tests/equality/equality.cpp

```diff
@@ -0,0 +1,1034 @@
+    libcellml::VariablePtr v2 = libcellml::Variable::create("vB");
+
+    c1->addVariable(v1);
+    c2->addVariable(v2);
+
+    EXPECT_FALSE(c1->equal(c2));
+    EXPECT_FALSE(c2->equal(c1));
+}
+
+TEST(Equality, importSourceEqual)
+{
+    libcellml::ImportSourcePtr is1 = libcellml::ImportSource::create();
+    libcellml::ImportSourcePtr is2 = libcellml::ImportSource::create();
+
+    is1->setUrl("great_url");
+    is2->setUrl("great_url");
+
+    EXPECT_TRUE(is1->equal(is2));
+    EXPECT_TRUE(is2->equal(is1));
+}
+
+TEST(Equality, importSourceNotEqual)
+{
+    libcellml::ImportSourcePtr is1 = libcellml::ImportSource::create();
+    libcellml::ImportSourcePtr is2 = libcellml::ImportSource::create();
+
+    is1->setUrl("another_great_url");
+    is2->setUrl("great_url");
+
+    EXPECT_FALSE(is1->equal(is2));
+    EXPECT_FALSE(is2->equal(is1));
+}
+
+TEST(Equality, importSourceNotEqualInvalidInput)
+{
+    libcellml::ImportSourcePtr is1 = libcellml::ImportSource::create();
+
+    is1->setUrl("a_great_url");
+
+    EXPECT_FALSE(is1->equal(nullptr));
+}
+
+TEST(Equality, modelEqual)
+{
+    libcellml::ModelPtr m1 = libcellml::Model::create();
+    libcellml::ModelPtr m2 = libcellml::Model::create();
+
+    libcellml::ComponentPtr c1 = libcellml::Component::create("c");
+    libcellml::ComponentPtr c2 = libcellml::Component::create("c");
+
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsA");
+
+    m1->addComponent(c1);
+    m2->addComponent(c2);
+
+    m1->addUnits(u1);
+    m2->addUnits(u2);
+
+    EXPECT_TRUE(m1->equal(m2));
+    EXPECT_TRUE(m2->equal(m1));
+}
+
+TEST(Equality, modelNotEqualInvalidInput)
+{
+    libcellml::ModelPtr m1 = libcellml::Model::create();
+
+    EXPECT_FALSE(m1->equal(nullptr));
+}
+
+TEST(Equality, modelNotEqualByComponents)
+{
+    libcellml::ModelPtr m1 = libcellml::Model::create();
+    libcellml::ModelPtr m2 = libcellml::Model::create();
+
+    libcellml::ComponentPtr c1 = libcellml::Component::create("c1");
+    libcellml::ComponentPtr c2 = libcellml::Component::create("c2");
+
+    m1->addComponent(c1);
+    m2->addComponent(c2);
+
+    EXPECT_FALSE(m1->equal(m2));
+    EXPECT_FALSE(m2->equal(m1));
+}
+
+TEST(Equality, modelNotEqualByUnits)
+{
+    libcellml::ModelPtr m1 = libcellml::Model::create();
+    libcellml::ModelPtr m2 = libcellml::Model::create();
+
+    libcellml::UnitsPtr u1 = libcellml::Units::create("unitsA");
+    libcellml::UnitsPtr u2 = libcellml::Units::create("unitsB");
+
+    m1->addUnits(u1);
+    m2->addUnits(u2);
+
+    EXPECT_FALSE(m1->equal(m2));
+    EXPECT_FALSE(m2->equal(m1));
+}
+
+TEST(Equality, modelHH)
+{
+    libcellml::ParserPtr parser = libcellml::Parser::create();
+    libcellml::ModelPtr model = parser->parseModel(fileContents("generator/hodgkin_huxley_squid_axon_model_1952/model.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
```

Comment:
Somewhere here I'd expect a test that looks at equality between the different ways of creating a model:
- parsing
- through the API entirely
- cloning
- flattening
I suspect that this will show that we don't have good things happening in the MathML strings, since the Parser (I think?) does some reformatting, which is different to how the maths is stored if you simply add it to a component?  
In any case, for completeness I think these tests are needed.

Reply:
945d4ec

Reply:
just to confirm that flattening isn't a problem, perhaps add a test like

```
auto parser = libcellml::Parser::create();
auto model = parser->parseModel(fileContents("modelflattening/outsidecomponentequivalentvariables.xml"));
auto clonedModel = model->clone();
auto importer = libcellml::Importer::create();
importer->resolveImports(model, resourcePath("modelflattening/"));
model = importer->flattenModel(model);

importer->resolveImports(clonedModel,  resourcePath("modelflattening/"));
clonedModel = importer->flattenModel(clonedModel);

EXPECT_TRUE(model->equals(clonedModel));
EXPECT_TRUE(clonedModel->equals(model));
```

Reply:
ef8426a

Reply:
I'm good to resolve this one, but just want to check @kerimoyle is also good.